### ISSUE TITLE
DOI and URN plugins allow undesirable changes 

### ISSUE THREAD

**Issue by mtub** [pkp/pkp-lib#408]
The existing pubId plugins (DOI, URN) seem to be constructed with a test scenario in mind, not with running them in a live environment. It's too easy to change, delete, or reassign identifiers without any check if they have already been registered (in which case these changes shouldn't be allowed).

Especially reassigning new identifiers after changing the template for generating identifiers shouldn't be possible at all.

If this is needed for testing purposes, there should at least exist a debug switch or a test if the pubId plugin is running with a test configuration or is registering real identifiers.

**Comment by asmecher** [comment]
Marco, in recent releases of OJS, the identifiers get attached to the articles when they're first served to a user in the reader front-end; thereafter they shouldn't change, even if the patterns in journal setup are altered. Is that different from the behavior you're seeing? (One source of confusion is the point at which the identifiers are attached: it's only when they're first presented to a reader.)

**Comment by bozana** [comment]
I think Marco means above all the "Reassing DOIs" button in the DOI plug-in settings and that this button should only be available in the testing mode, but not in the live/publishing mode :-)

**Comment by asmecher** [comment]
If that's the case: to be fair, we do hide some very apocalyptic options behind an "are you sure?" confirmation modal. Thinking e.g. of deleting a journal.

**Comment by bozana** [comment]
:-) Yes, this is also here the case... :-)

**Comment by mtub** [comment]
I think that (number of people knowing that deleting a journal might be bad) >> (number of people knowing that reassigning already registered DOI is probably bad).

And to push this a little further: The language used around this option is misleading. The text reads:

>If you change your DOI configuration, DOIs that have already been assigned will not be affected. Once the DOI configuration is saved, use this button to clear all existing DOIs so that the new settings will take effect with existing articles.

So what happens is that existing (already registered) DOI are deleted (but only within OJS's db?) in order to be able to re-generate identifiers. That option should maybe be named "delete DOI" or "delete and reassign DOI", but not just reassign. But when dealing with a live environment, this option shouldn't be available at all.

I still stand by what I said in the issue description: It's too easy to consider permanent identifiers as something to play around with, like changing the pattern for generating DOIs and then rebuild them for articles that have already been registered with crossref, datacite, etc.

**Comment by mtub** [comment]
I would like to push this and repeat that the app configuration should not allow deleting assigned DOIs. Requirements of a test mode should not take precedence over real-world requirements. Protecting the persistence of DOIs is an important requirement.

**Comment by AhemNason** [comment]
Hi, 

So,  in some ways, this has been half addressed in Bozana's more recent DOI/Crossref work. There's now an "assign DOIs" button as well as a "reassign DOIs" button. The "assign DOIs" button is great because it makes a DOI for anything that doesn't currently have one. I won't lie in stating that there's usefulness to these buttons for me and in general, but their use is not obvious to regular users. 

Here's a use case example. 

Currently, if you turn on DOIs but do not assign them, OJS (at least 2) will assign a DOI to an article that doesn't have one when someone loads the record. This is sort of weird behaviour that I haven't seen past 3.0, but basically, there's no harm in changing a DOI _before the DOI is registered_. 

It's not unusual for a DOI to be assigned to something accidentally that a user would like to change before "locking it in", so to speak. This is especially the case if the user turned on automatic DOI assignment in the early days of setting up a journal but didn't even know they had to register them (this is astoundingly common). 

In fact, I actually sort of think (assuming this is possible), that no DOI should even appear with a record until it's been successfully registered with an agency. If it doesn't resolve, it probably shouldn't display on the article view. 

But, I _very much do think_ that once a DOI has been registered with Crossref (or whomever) that changing that DOI in OJS should be something only an administrator can do, possibly only via command line. 

There are a handful of admin-only options I think should exist with the DOI plugin. One should be batch XML export of whole swaths of issues at once. The other is this "reassigning" of DOIs

**Comment by mtub** [comment]
Thanks, @AhemNason. Regarding your last point: Once a DOI has been registered, what is the use case for re-assigning another DOI and wiping the already registered one from the database? 

Every example I have been presented with where someone needed this option was for reasons that we should not support, e.g. aligning DOIs after getting a new prefix (change of publisher…), thinking of a fancier prefix, etc. I don't think the risk is editors misunderstanding the option, I think the risk is editors knowing what it does and using it for the wrong reasons. And editors would also be able to tell an admin to just do it.

But I agree that the risk of doing harm is smaller if this was an admin-only option.

**Comment by mtub** [comment]
In general, the problem with the pubId plugins is that we don't know if an identifier has been registered. We could enhance the plugins to check if an identifier resolves, but this would also result in warnings when a resolver service is temporarily down. But at least it would enable us to say "if it resolves correctly, don't touch it"

**Comment by AhemNason** [comment]
There is rarely a use-case for assigning another DOI and wiping an already registered one. Or, if there is, it's _very very specific_. 

But I should make it clear that assignment of a DOI and registration of a DOI are very different things. If I publish an issue and OJS mints DOIs for my whole issue, but I haven't successfully registered them with Crossref (or whomever), then changing them isn't an issue at all. 

(well, that's sort of debatable, actually, because if I publish an issue and those unregistered DOIs proliferate, then you might end up with conflicts if you change them and register different ones. It really depends on how much time passes.)

Right now, if you mint DOIs and they fail to register (let's say you're missing a piece of metadata needed and that registration fails), then those DOIs aren't "locked in". 

I think one thing I genuinely believe when it comes to most editors and DOIs is that they really don't have a huge grasp on what they're for or how they work. I'm not especially concerned that editors will ask admins to mint a new batch, and Crossref are very transparent that you shouldn't mint two DOIs for one article. They'll actually charge you to fix the issue if it's prolific enough. 

Basically, the only situation I can think of where I'd need to remint every DOI would be if a journal left one publisher to go to another, and the new publisher (or the journal itself) could not, for whatever reason, wrestle ownership of the old DOIs away from that body. Currently, if you move publishers, you have to contact Crossref, they communicate with the publisher, and move ownership of those DOIs to the new body (because the prefix doesn't change, but it belongs to another registrant). If you weren't able to do that for whatever reason, you'd have to remint your DOIs. That's a pretty big edge case.

**Comment by AhemNason** [comment]
I'd definitely advocate for anything that prevented a DOI from changing once it had been successfully registered with an agency, though. That would be very good.

**Comment by ewhanson** [comment]
@asmecher, yes, this is no longer an issue for DOIs. They are now locked from editing once they have been registered. This does not affect URNs, though, so any issue that was present in 3.3 for URNs will still be present now.

**Comment by NateWr** [comment]
Closing this as outdated. If you feel this is still important, please consider making a proposal in the [feature request category](https://forum.pkp.sfu.ca/c/feature-requests/8) of our community forum.
